bsip-0040.md

```diff
@@ -0,0 +1,322 @@
+`data = [ [restriction 1, restriction 2], [restriction 3, restriction 4], ... ]`. If one of the restrictions sub-lists in data passes as whole, this restriction passes. In above miniexample that would mean if `restriction 1` and `restriction 2` pass, the whole `logical_or` restriction is considered to be passed as well.
+
+### Outline of handling incoming transactions
+ 
+When a signed transaction arrives and before the backend evaluates if all necessary authorities are present through the signatures, do the following:
+- iterate over required accounts and for each account, iterate over all operations within the transactions that require the active authority of this account
+- iterate the `custom_active_authorities` of said account
+- if a `custom_active_authority` is found that matches , remember that and stop iterating the authorities and continue until all operations are checked
+- if the account has a `custom active authority` match for every operation in the transaction that requires it, then grant the `active authority` of said account. If no match is found, treat as if no authority was given
+
+Note:
+- A `custom_active_authority` can only grant the `active authority` of the corresponding account, nothing more
+
+### Economics
+
+Adding a custom active authority means increased effort for the backend, and with a stateful one also the need for more storage. Proposed transaction fees:
+- `install_custom_active_authority`: Tied to the duration of the custom active authority. 
+Normal accounts can only create custom active authoritites with a duration of maximum 1 year. LTM can do any duration and also unlimited, but the transaction fee is capped at duration of 2 years.
+- `update_custom_active_authority`: Base fee similar to `account_update` plus dynamic one for any duration changes
+- `delete_custom_active_authority`: Cheap similar to `limit_order_cancel`
+
+### Modification to the backend
+
+* Add a new index or extend the account object to store custom active permission are assigned to an account and contain a list of custom active authorities. Multiple custom active authority entries are possible for one operation
+* If the active authority of the account is updated, all custom active authorities need to be confirmed in the update. Every unconfirmed one is deleted otherwise
+* Provide operations: `install_custom_active_authority`, `update_custom_active_authority`, `delete_custom_active_authority` to allow changing the custom active permission (3 operation to allow custom transaction fees and avoid having to send the complete list of all authorities for every update)
+* Operation-specific authorities (if present) must be evaluated in incoming transactions
+* Additional committee parameters may be needed to limit the extend of usage of this feature
+
+Notes: The implementation must not differentiate on which operation the custom active authority is applied, all operations are treated in same fashion
```

Comment:
Are you sure? What about proposal_update? :-)

Reply:
The only thing a custom_operation_permission for proposal_update could do is to add the `active` permission of an account to a pending proposal (or remove it).

This could serve as a true 2FA/Multisig setup where the other party has even less power over the funds than in a regular multisig setup.

Reply:
That is true, but intentional. Such an "approval key" that allows `proposal_update` could approve anything, but can't create proposals. Like Fabian mentioned that enables generic 2FA, where I would have such a "approval key" e.g. on my cellphone.

Reply:
Please be aware that everyone can create proposals for you. So if that key is compromised, effectively you can lose all funds in your account. You can still change permissions with your owner authority though.